### PR TITLE
Add Ability to perform Node operation on IBM Cloud IPI deployment

### PR DESCRIPTION
**Pull request by prsurve**

### CODE REVIEW

File: ocs_ci/utility/ibmcloud.py [hunk 1]
```diff
@@ -520,6 +522,268 @@ def delete_volume_id(self, volume):
...
+        self.stop_nodes(nodes=nodes, force=force)
+        if wait:
+            for node in nodes:
+                sample = TimeoutSampler(
+                    timeout=timeout,
+                    sleep=10,
+                    func=self.check_node_status,
+                    node_name=node.name,
+                )
+                sample.wait_for_func_status(result=constants.STATUS_STOPPED)
```
Comment (yitzhak12) [inline ocs_ci/utility/ibmcloud.py:608]: The `result` param should be a boolean value(See the function definition here: https://github.com/prsurve/ocs-ci/blob/6caf2667833e887f9df7392af3666067aa6493ee/ocs_ci/utility/utils.py#L1342). Maybe you can create a new method `is_node_running`, and use it here instead.

Comment (yitzhak12) [reply ocs_ci/utility/ibmcloud.py:608]: Same thing for the other occurences.

File: ocs_ci/utility/ibmcloud.py [hunk 2]
```diff
@@ -520,6 +522,268 @@ def delete_volume_id(self, volume):
...
+        """
+        try:
+            cmd = f"ibmcloud is instance {node_name} --output json"
+
+            out = run_ibmcloud_cmd(cmd)
+            out = json.loads(out)
+            return out["status"]
+        except CommandFailed as cf:
+            if "Instance not found" in str(cf):
+                return True
```
Comment (yitzhak12) [inline ocs_ci/utility/ibmcloud.py:638]: Why here do we return a boolean value if the method should return `str`?

File: ocs_ci/utility/ibmcloud.py [hunk 3]
```diff
@@ -520,6 +522,268 @@ def delete_volume_id(self, volume):
...
+        assert resource_name, "Resource Not found"
+        cmd = f"ibmcloud is ins --resource-group-name {resource_name} --output json"
+        out = run_ibmcloud_cmd(cmd)
+        all_instance_output = json.loads(out)
+        for instance_name in all_instance_output:
+            if instance_name["status"] == constants.STATUS_STOPPED:
+                node_obj = OCP(kind="Node", resource_name=instance_name["name"]).get()
+                node_obj_ocs = OCS(**node_obj)
+                stop_node_list.append(node_obj_ocs)
+            if instance_name["status"] == constants.STATUS_STOPPED:
```
Comment (yitzhak12) [inline ocs_ci/utility/ibmcloud.py:665]: I think it should be `if instance_name["status"] == constants.STATUS_STOPPING:`.

File: ocs_ci/utility/ibmcloud.py [hunk 4]
```diff
@@ -520,6 +522,268 @@ def delete_volume_id(self, volume):
...
+                node_obj = OCP(kind="Node", resource_name=instance_name["name"]).get()
+                node_obj_ocs = OCS(**node_obj)
+                stopping_node_list.append(node_obj_ocs)
+                stop_node_list.append(node_obj_ocs)
+        logger.info("Force stopping node which are in stopping state")
+        self.stop_nodes(nodes=stop_node_list, force=True, wait=True)
```
Comment (yitzhak12) [inline ocs_ci/utility/ibmcloud.py:671]: I think it should be `self.stop_nodes(nodes=stopping_node_list, force=True, wait=True)`

File: ocs_ci/utility/ibmcloud.py [hunk 5]
```diff
@@ -520,6 +522,268 @@ def delete_volume_id(self, volume):
...
+            )
+            out = run_ibmcloud_cmd(cmd)
+            logger.info(f"detachment command output: {out}")
+
+    def attach_volume(self, volume, node):
+        """
+        Attach volume to node on IBM Cloud.
+        Args:
+            volume (str): volume id.
+            node (OCS): worker node id to attach.
```
Comment (yitzhak12) [inline ocs_ci/utility/ibmcloud.py:734]: Typo fix. It should be `worker node object to detach.`

File: ocs_ci/utility/ibmcloud.py [hunk 6]
```diff
@@ -520,6 +522,268 @@ def delete_volume_id(self, volume):
...
+        """
+        logger.info(
+            f"attach_volumes:{node[0].get()['metadata']['labels']['failure-domain.beta.kubernetes.io/zone']}"
```
Comment (yitzhak12) [inline ocs_ci/utility/ibmcloud.py:737]: I think `node[0]` should be replaced with `node`.

File: ocs_ci/utility/ibmcloud.py [hunk 7]
```diff
@@ -520,6 +522,268 @@ def delete_volume_id(self, volume):
...
+        logger.info(f"volume is : {volume}")
+
+        cmd = f"ibmcloud is volume {volume} --output json"
+        out = run_ibmcloud_cmd(cmd)
+        out = json.loads(out)
+
+        if len(out["volume_attachments"]) == 0:
+
+            logger.info(f"attachment command output: {out}")
+            cmd = f"ibmcloud is instance-volume-attachment-add data-vol-name {node[0].name} {volume} --output json"
```
Comment (yitzhak12) [inline ocs_ci/utility/ibmcloud.py:749]: I think `node[0]` should be replaced with `node`.

File: ocs_ci/utility/ibmcloud.py (hunk 1 above)
Comment (prsurve) [reply ocs_ci/utility/ibmcloud.py]: i have update the code

File: ocs_ci/utility/ibmcloud.py (hunk 4 above)
Comment (prsurve) [reply ocs_ci/utility/ibmcloud.py:679]: i have update the code

File: ocs_ci/utility/ibmcloud.py (hunk 5 above)
Comment (prsurve) [reply ocs_ci/utility/ibmcloud.py]: i have update the code

File: ocs_ci/utility/ibmcloud.py (hunk 3 above)
Comment (prsurve) [reply ocs_ci/utility/ibmcloud.py:674]: There are only 2 state so i'm using Stopped

File: ocs_ci/utility/ibmcloud.py [hunk 8]
```diff
@@ -520,6 +522,268 @@ def delete_volume_id(self, volume):
...
+                )
+                sample.wait_for_func_status(result=True)
+                break
+
+    def detach_volume(self, volume, node=None):
+        """
+        Detach volume from node on IBM Cloud.
+        Args:
+            volume (str): volume id.
+            node (OCS): worker node id to detach.
```
[the remark this answers is not in the document]
Comment (prsurve) [reply ocs_ci/utility/ibmcloud.py]: i have update the code

File: ocs_ci/utility/ibmcloud.py (hunk 6 above)
Comment (prsurve) [reply ocs_ci/utility/ibmcloud.py]: i have update the code

File: ocs_ci/utility/ibmcloud.py (hunk 7 above)
Comment (prsurve) [reply ocs_ci/utility/ibmcloud.py]: i have update the code

File: ocs_ci/utility/ibmcloud.py [hunk 9]
```diff
@@ -520,6 +522,268 @@ def delete_volume_id(self, volume):
...
+        Args:
+            volume (str): The volume to check for to attached
+        Returns:
+            bool: 'True' if volume is attached otherwise 'False'
+        """
+        logger.info("Checking volume attachment status")
+        cmd = f"ibmcloud is volume {volume} --output json"
+        out = run_ibmcloud_cmd(cmd)
+        out = json.loads(out)
+        return out["volume_attachments"]
```
[the remark this answers is not in the document]
Comment (prsurve) [reply ocs_ci/utility/ibmcloud.py]: i have update the code

File: ocs_ci/utility/ibmcloud.py (hunk 3 above)
Comment (yitzhak12) [reply ocs_ci/utility/ibmcloud.py:674]: Don't we want to start the nodes in status `stop` and `stopping`? and if not, we should remove the code: 

> if instance_name["status"] == constants.STATUS_STOPPED:
                node_obj = OCP(kind="Node", resource_name=instance_name["name"]).get()
                node_obj_ocs = OCS(**node_obj)
                stop_node_list.append(node_obj_ocs)

as duplicate.

File: ocs_ci/utility/ibmcloud.py [hunk 10]
```diff
@@ -520,6 +522,280 @@ def delete_volume_id(self, volume):
...
+                )
+                sample.wait_for_func_status(result=True)
+
+    def check_node_status(self, node_name, node_status):
+        """
+        Check the node status in IBM cloud
+
+        Args:
+            node_name (str): Node name
+
```
Comment (yitzhak12) [inline ocs_ci/utility/ibmcloud.py:632]: Please add the new param `node_status` in the docstring.

File: ocs_ci/utility/ibmcloud.py [hunk 11]
```diff
@@ -520,6 +522,280 @@ def delete_volume_id(self, volume):
...
+        Returns:
+            str: Status of node
```
Comment (yitzhak12) [inline ocs_ci/utility/ibmcloud.py:634]: Please change the return type to `bool` with the relevant description.

File: ocs_ci/utility/ibmcloud.py (hunk 10 above)
Comment (prsurve) [reply ocs_ci/utility/ibmcloud.py:633]: I have Updated the PR

**Review by yitzhak12 — DISMISSED** [review]

**Review by am-agrawa — DISMISSED** [review]

File: ocs_ci/ocs/platform_nodes.py [hunk 12]
```diff
@@ -3345,3 +3350,111 @@ def terminate_nodes(self, nodes, wait=True):
 
         """
         raise NotImplementedError("Terminate nodes functionality not implemented")
+
+
+class IBMCloudIPI(object):
+    """
+    A base class for nodes related operations.
```
[the remark this answers is not in the document]
Comment (prsurve) [reply ocs_ci/ocs/platform_nodes.py]: Done

File: ocs_ci/utility/ibmcloud.py [hunk 13]
```diff
@@ -574,6 +576,280 @@ def delete_volume_id(self, volume):
             logger.info("volume is not deleted")
 
 
+class IBMCloudIPI(object):
```
[the remark this answers is not in the document]
Comment (prsurve) [reply ocs_ci/utility/ibmcloud.py:579]: Done

**Review by yitzhak12 — APPROVED** [review]

**Review by amr1ta — APPROVED** [review]

**Review by ebenahar — APPROVED** [review]

**Pull request merged**